Refactor sponsor invoice update/notify system to avoid issues caused by `switch_to_blog`

coreymckrill opened the pull request:
Removes `switch_to_blog` wherever possible in the ecosystem of functions that processes and notifies about changes to sponsor invoices.

**To do:**

* [ ] ~Maybe make email message strings translatable that currently aren't~ This PR is complicated enough, we can do this later.
* [x] Add more temporary logging or email to devs for tracking/troubleshooting

Fixes #358 

**To test:**

1. Follow all the testing steps in #371 to create two or three invoices and check that they have all been sent over to the QBO sandbox. When you click the "Approve" button on an invoice in the network Invoices dashboard, you should be able to refresh the screen and see that it has moved from the "Submitted" tab to the "Sent" tab. Then also in the Sponsor Invoices list table over on the test site, you should see that the status of the invoice changes as well. (Success here indicates that both the post status and the entry in the index table are being updated correctly)
1. In the QBO sandbox company dashboard, on the Invoices screen, choose one of the invoices that you just created and click the button to "Receive payment"
1. Open up MailCatcher so you can see email notifications being sent from the test environment.
1. In order to test the cron job that checks for paid invoices, you can use wp-cli to immediately trigger a scheduled event. So first, test that running the cron job on a different site from the one where the invoices are does _not_ mark anything as paid:
    `wp --allow-root --url=https://wrongsite.wordcamp.test cron event run wcbdsi_check_for_paid_invoices`
1. Then go to the site that does have the sponsor invoices:
    `wp --allow-root --url=https://rightsite.wordcamp.test cron event run wcbdsi_check_for_paid_invoices`
1. The invoice that received payment should move from "Sent" to "Paid" in the network Invoices dashboard, and the status of the invoice post in the test site should also update to Paid. MailCatcher should have an email about the invoice being paid. Note that this email notification is not translatable, so it will be in `en_US`.
1. In order to test the pending invoice reminder email cron job, you'll need to add a temporary hack to the code. In the `send_invoice_pending_reminder()` function, comment out lines 551 through 561, so it will send for all pending invoices, even though you just created them today, and also not mark them as defaulted in the case that your test WordCamp start date is already in the past.
1. Change the locale to something other than `en_US` (to test whether this email notification, which is translatable, is sent using the correct locale).
1. As with the above tests for finding paid invoices, try first on the wrong site to make sure no reminder emails are sent, and then from the correct site to make sure the emails _are_ sent, and in the correct locale:
    `wp --allow-root --url=https://wrongsite.wordcamp.test cron event run send_invoice_pending_reminder`
    `wp --allow-root --url=https://rightsite.wordcamp.test cron event run send_invoice_pending_reminder`
1. When finished, don't forget to remove the temporary hack 😄

coreymckrill committed:
Refactor `send_invoice_pending_reminder()`

Only pull pending invoices from the current site, instead of all
pending invoices in the index table. This way the notifications about
those invoices will appear to come from the correct WordCamp site,
and be in the correct language.

The cron job is set up to run on every site in the network, so this
shouldn't cause any reminders to get missed.

coreymckrill committed:
Add developer address to email notifications

Temporarily add the developer email address as a recipient on the
pending invoice notifications so we can track whether these are
working correctly.

coreymckrill committed:
phpcs

coreymckrill committed:
Reorganize scheduling of cron jobs

Put all the event scheduling in one place, and make sure it runs
consistently, instead of only on network admin or cron requests.

coreymckrill committed:
Refactor more invoice processing functions

* `check_for_paid_invoices`: Another cron job that only needs to process
  invoices from the current site, since it runs on every site.
* `mark_invoices_as_paid`, `update_invoice_status`,
  `notify_organizer_status_changed`: Functions used by
  `check_for_paid_invoices` that no longer need to switch blogs.
* `handle_approve_invoice_request`: This one is an Ajax handler that
  still needs switch blogs since it is executed from the network admin.
  However its usage of `update_invoice_status` and
  `notify_organizer_status_changed` still needed to change.

coreymckrill committed:
Add post type checks for things that assume a sponsor invoice post type

coreymckrill committed:
phpcs

iandunn commented on public_html/wp-content/plugins/wordcamp-payments-network/includes/sponsor-invoices-dashboard.php:
```diff
@@ -481,32 +483,29 @@ function schedule_sent_invoice_reminder() {
+ *
+ * This runs as a cron job on every site, so it only needs to look for invoices that are from the current site.
+ * (Previously it pulled all pending invoices from the index table, which caused weird issues like the email about
+ * an invoice in Chicago coming from a WordCamp site in Germany.)
```
```suggestion
 * an invoice in Chicago coming from a WordCamp site in Germany, translated in `de_DE`.)
```

iandunn reviewed (commented):
The code itself LGTM 👍 

You mentioned some testing problems on Slack, how did you end up testing it?

Beyond verifying that the emails are sent w/ the expected "From" and locale, one side-effect I'd check would be to make sure the index table is kept in sync w/ the site's `wp_N_posts` table.

coreymckrill committed:
Use correct meta key for QBO invoice ID

coreymckrill committed:
Move index row actions outside of state conditionals

There's no need to limit which states (DOING_CRON etc) to fire these
actions for, and it just introduces the possibility of the actions
unexpectedly _not_ firing.

coreymckrill committed:
Update public_html/wp-content/plugins/wordcamp-payments-network/includes/sponsor-invoices-dashboard.php

Co-Authored-By: Ian Dunn <ian@iandunn.name>

coreymckrill commented:
This now has testing instructions since #371 has been merged.

> Beyond verifying that the emails are sent w/ the expected "From" and locale, one side-effect I'd check would be to make sure the index table is kept in sync w/ the site's wp_N_posts table.

This is a good point. The instructions include verification of this.
